### PR TITLE
[Algorithm] Support H2RBox

### PR DESCRIPTION
**Pull request by yangxue0827**
## Motivation

- **Support weakly supervised oriented object detection algorithm [H2Rbox](https://arxiv.org/abs/2210.06742) in MMRotate 1.x**
- **Support DOTA-v1.5 and DOTA-v2.0**

## Modification

- Create `h2rbox.py` at [detectors](https://github.com/open-mmlab/mmrotate/tree/1.x/mmrotate/models/detectors)
- Create `h2rbox_head.py` at [dense_heads](https://github.com/open-mmlab/mmrotate/tree/1.x/mmrotate/models/dense_heads)
- Create `DOTAv15Dataset` and `DOTAv2Dataset` in [dota.py](https://github.com/open-mmlab/mmrotate/tree/1.x/mmrotate/datasets/dota.py)
- Create `dotav15.py` and `dotav2.py` at [configs/_base_/datasets](https://github.com/open-mmlab/mmrotate/tree/1.x/configs/_base_/datasets)
- Create `h2rbox_consistency_loss.py` at [losses](https://github.com/open-mmlab/mmrotate/tree/1.x/mmrotate/models/losses)
- Create `README.md` and  `metafile.yml` at [configs/h2rbox](https://github.com/open-mmlab/mmrotate/tree/1.x/configs)
- Create `h2rbox-le90_r50_fpn_adamw-1x_dota.py` , `h2rbox-le90_r50_fpn_adamw-3x_dota.py` and `h2rbox-le90_r50_fpn_adamw-1x_dota-ms.py` at [configs/h2rbox](https://github.com/open-mmlab/mmrotate/tree/1.x/configs)
- Create `h2rbox-le90_r50_fpn_adamw-1x_dotav15.py` and `h2rbox-le90_r50_fpn_adamw-3x_dotav15.py` at [configs/h2rbox/dotav15](https://github.com/open-mmlab/mmrotate/tree/1.x/configs)
- Create `h2rbox-le90_r50_fpn_adamw-1x_dotav2.py` and `h2rbox-le90_r50_fpn_adamw-3x_dotav2.py` at [configs/h2rbox/dotav2](https://github.com/open-mmlab/mmrotate/tree/1.x/configs)
- Create `h2rbox-le90_r50_fpn_adamw-1x_dior.py` at [configs/h2rbox/dior](https://github.com/open-mmlab/mmrotate/tree/1.x/configs)
- Create `test_h2rbox` at [test_detectors](https://github.com/open-mmlab/mmrotate/tree/1.x/tests/test_models/test_detectors)
- Update `test_loss.py` at [test_losses](https://github.com/open-mmlab/mmrotate/tree/1.x/tests/test_models/test_losses)
- Update `test_dota.py` at [test_datasets](https://github.com/open-mmlab/mmrotate/tree/1.x/tests/test_datasets)
- Update `test_h2rbox_heads.py` at [test_dense_heads](https://github.com/open-mmlab/mmrotate/tree/1.x/tests/test_models/test_dense_heads)

## Performance

- **Performance on DOTA1.0**

This PR
|         Backbone         |  mAP  | Angle | lr schd | Mem (GB) | Inf Time (fps) | MS  | Batch Size |                                     Configs                                     |   
|:------------------------:|:-----:|:-----:|:-------:|:--------:|:--------------:|:---:|:----------:|:-------------------------------------------------------------------------------:|
| ResNet50 (1024,1024,200) | **68.75** | le90  |   1x    |    6.25      |                |  -  |     2      |   h2rbox-le90_r50_fpn_adamw-1x_dota   | 
| ResNet50 (1024,1024,200) | **71.31** | le90  |   3x    |     6.64     |                |  -  |     2      | h2rbox-le90_r50_fpn_adamw-3x_dota |
| ResNet50 (1024,1024,200) | **74.43** | le90  |   1x    |     6.12     |        -        |  -  |     2      | h2rbox-le90_r50_fpn_adamw-1x_dota-ms |

[Original PyTorch Implementation (Paper)](https://github.com/yangxue0827/h2rbox-mmrotate)

|         Backbone         |  mAP  | Angle | lr schd | Mem (GB) | Inf Time (fps) | MS  | Batch Size |                             Configs                                                         |
|:------------------------:|:-----:|:-----:|:-------:|:--------:|:--------------:|:---:|:----------:|:-------------------------------------------------------------------------------------------:|
| ResNet50 (1024,1024,200) | 67.45 | le90  |   1x    |   7.02   |      28.5      |  -  |     2      |         h2rbox_r50_adamw_fpn_1x_dota_le90         |
| ResNet50 (1024,1024,200) | 70.77 | le90  |   3x    |   7.02   |      28.5      |  -  |     2      |                   h2rbox_r50_adamw_fpn_3x_dota_le90                  | 
| ResNet50 (1024,1024,200) | 74.40 | le90  |   1x    |   8.58   |       -     |  -  |     2      |                   h2rbox_r50_adamw_fpn_3x_dota_ms_le90                  | 

[Jittor Implementation](https://github.com/yangxue0827/h2rbox-jittor)

|         Backbone         |  mAP  | Angle | lr schd | MS  | Batch Size |                                                      Configs                                                      |
|:------------------------:|:-----:|:-----:|:-------:|:---:|:----------:|:-----------------------------------------------------------------------------------------------------------------:|
| ResNet50 (1024,1024,200) | 67.62 | le135 |   1x    |  -  |     2      |                 h2rbox_obb_r50_adamw_fpn_1x_dota                 | 

- **Performance on DOTA1.5**

This PR
|         Backbone         |  mAP  | Angle | lr schd | Mem (GB) | Inf Time (fps) | MS  | Batch Size |                                     Configs                                     |   
|:------------------------:|:-----:|:-----:|:-------:|:--------:|:--------------:|:---:|:----------:|:-------------------------------------------------------------------------------:|
| ResNet50 (1024,1024,200) | **60.33** | le90  |   1x    |     6.83     |                |  -  |     2      |   h2rbox-le90_r50_fpn_adamw-1x_dotav15   | 
| ResNet50 (1024,1024,200) | **62.65** | le90  |   1x    |   6.34   |                |  -  |     2      | h2rbox-le90_r50_fpn_adamw-3x_dotav15 |

[Original PyTorch Implementation (Paper)](https://github.com/yangxue0827/h2rbox-mmrotate)

|         Backbone         |  mAP  | Angle | lr schd | Mem (GB) | Inf Time (fps) | MS  | Batch Size |                             Configs                                                         |
|:------------------------:|:-----:|:-----:|:-------:|:--------:|:--------------:|:---:|:----------:|:-------------------------------------------------------------------------------------------:|
| ResNet50 (1024,1024,200) | 60.19 | le90  |   1x    |   10.68   |      25.8      |  -  |     2      |         h2rbox_r50_adamw_fpn_1x_dotav15_le90         |
| ResNet50 (1024,1024,200) | 62.60 | le90  |   3x    |   10.68   |      25.8      |  -  |     2      |         h2rbox_r50_adamw_fpn_3x_dotav15_le90         |

- **Performance on DOTA2.0**

This PR
|         Backbone         |  mAP  | Angle | lr schd | Mem (GB) | Inf Time (fps) | MS  | Batch Size |                                     Configs                                     |   
|:------------------------:|:-----:|:-----:|:-------:|:--------:|:--------------:|:---:|:----------:|:-------------------------------------------------------------------------------:|
| ResNet50 (1024,1024,200) | **47.08** | le90  |   1x    |     7.58     |                |  -  |     2      |   h2rbox-le90_r50_fpn_adamw-1x_dotav2   |
| ResNet50 (1024,1024,200) | **50.20** | le90  |   1x    |     7.74     |                |  -  |     2      |   h2rbox-le90_r50_fpn_adamw-3x_dotav2   | 

[Original PyTorch Implementation (Paper)](https://github.com/yangxue0827/h2rbox-mmrotate)

|         Backbone         |  mAP  | Angle | lr schd | Mem (GB) | Inf Time (fps) | MS  | Batch Size |                             Configs                                                         |
|:------------------------:|:-----:|:-----:|:-------:|:--------:|:--------------:|:---:|:----------:|:-------------------------------------------------------------------------------------------:|
| ResNet50 (1024,1024,200) |  45.87 | le90  |   1x    |   11.57   |      25.0      |  -  |     2      |         h2rbox_r50_adamw_fpn_1x_dotav2_le90         |
| ResNet50 (1024,1024,200) |  47.86 | le90  |   3x    |   11.57   |      25.0      |  -  |     2      |         h2rbox_r50_adamw_fpn_3x_dotav2_le90         |

- **Performance on DIOR**

This PR
|         Backbone         |  AP50:95 | AP50 | AP75  | Angle | lr schd | Mem (GB) | Inf Time (fps) | MS  | Batch Size |                                        Configs                                         | 
| :----------------------: | :---: | :---: | :---: | :---: | :-----: | :------: | :------------: | :-: | :--------: | :------------------------------------------------------------------------------------: |
| ResNet50 (1024,1024,200) | 33.06 | **57.40** | 32.50 | le90  |   1x    |   3.83   |                |  -  |     2      | h2rbox-le90_r50_fpn_adamw-1x_dior |

[Original PyTorch Implementation (Paper)](https://github.com/yangxue0827/h2rbox-mmrotate)
|         Backbone         |  AP50:95 | AP50 | AP75  | Angle | lr schd | Mem (GB) | Inf Time (fps) | MS  | Batch Size |                                        Configs                                         | 
| :----------------------: | :---: | :---: | :---: | :---: | :-----: | :------: | :------------: | :-: | :--------: | :------------------------------------------------------------------------------------: |
| ResNet50 (1024,1024,200) | **33.15** | 57.00 | **32.60** | le90  |   1x    |   3.83   |                |  -  |     2      | h2rbox_r50_adamw_fpn_1x_dior_le90 |

## Notes
- `MS` means multiple scale image split.
- `Inf Time` was tested on a single RTX3090.
- The memory log of MMRotate 1.x is dynamic, so the last iteration is recorded here.

### CODE REVIEW

File: docs/zh_cn/model_zoo.md [hunk 1]
```diff
@@ -67,6 +68,7 @@
...
+|  ResNet50 (1024,1024,200)  | 68.75 |     le90     |    1x    |     6.25      |                |    -     |    2     |                          [h2rbox-le90_r50_fpn_adamw-1x_dota](../../configs/h2rbox/h2rbox-le90_r50_fpn_adamw-1x_dota.py)                           |                                                                                                                                                                                                   [model](<>) \\                                                                                                                                                                                                   |
```
Comment (zytx121) [inline docs/zh_cn/model_zoo.md:71]: update link

File: docs/en/model_zoo.md [hunk 2]
```diff
@@ -65,6 +66,7 @@
...
+|  ResNet50 (1024,1024,200)  | 68.75 | le90  |   1x    |   6.25   |                |   -   |     2      |                          [h2rbox-le90_r50_fpn_adamw-1x_dota](../../configs/h2rbox/h2rbox-le90_r50_fpn_adamw-1x_dota.py)                           |                                                                                                                                                                                                   [model](<>) \\                                                                                                                                                                                                   |
```
Comment (zytx121) [inline docs/en/model_zoo.md:69]: update link

File: mmrotate/models/dense_heads/h2rbox_head.py [hunk 3]
```diff
@@ -0,0 +1,710 @@
...
+from mmrotate.models.dense_heads.rotated_fcos_head import RotatedFCOSHead
+from mmrotate.registry import MODELS
+from mmrotate.structures import RotatedBoxes, hbox2rbox, rbox2hbox
+
+INF = 1e8
+
+
+@MODELS.register_module()
+class H2RBoxHead(RotatedFCOSHead):
+    """Anchor-free head used in `FCOS <https://arxiv.org/abs/1904.01355>`_.
```
Comment (zytx121) [inline mmrotate/models/dense_heads/h2rbox_head.py:26]: update paper link

File: mmrotate/models/dense_heads/h2rbox_head.py [hunk 4]
```diff
@@ -0,0 +1,710 @@
...
+        bbox_coder (:obj:`ConfigDict` or dict): Config of bbox coder. Defaults
+            to 'DistanceAnglePointCoder'.
+        loss_cls (:obj:`ConfigDict` or dict): Config of classification loss.
+        loss_bbox (:obj:`ConfigDict` or dict): Config of localization loss.
+        loss_centerness (:obj:`ConfigDict`, or dict): Config of centerness loss.
+        loss_angle (:obj:`ConfigDict` or dict, Optional): Config of angle loss.
+        loss_bbox_ss (:obj:`ConfigDict` or dict): Config of consistency loss.
+        rotation_agnostic_classes (list): Ids of rotation agnostic category.
+        weak_supervised (bool): If true, horizontal gtbox is input.
+            Default to True.
```
Comment (zytx121) [inline mmrotate/models/dense_heads/h2rbox_head.py:52]:
```suggestion
            Defaults to True.
```

File: mmrotate/models/dense_heads/h2rbox_head.py [hunk 5]
```diff
@@ -0,0 +1,710 @@
...
+    Compared with FCOS head, Rotated FCOS head add a angle branch to
+    support rotated object detection.
+
+    Args:
+        num_classes (int): Number of categories excluding the background
+            category.
+        in_channels (int): Number of channels in the input feature map.
+        angle_version (str): Angle representations. Defaults to 'le90'.
+        use_hbbox_loss (bool): If true, use horizontal bbox loss and
+            loss_angle should not be None. Default to False.
```
Comment (zytx121) [inline mmrotate/models/dense_heads/h2rbox_head.py:37]:
```suggestion
            loss_angle should not be None. Defaults to False.
```

File: mmrotate/models/dense_heads/h2rbox_head.py [hunk 6]
```diff
@@ -0,0 +1,710 @@
...
+        scale_angle (bool): If true, add scale to angle pred branch.
+            Default to True.
```
Comment (zytx121) [inline mmrotate/models/dense_heads/h2rbox_head.py:39]:
```suggestion
            Defaults to True.
```

File: mmrotate/models/dense_heads/h2rbox_head.py [hunk 7]
```diff
@@ -0,0 +1,710 @@
...
+                     use_sigmoid=True,
+                     loss_weight=1.0),
+                 loss_angle: OptConfigType = None,
+                 loss_bbox_ss: ConfigType = dict(
+                     type='mmdet.IoULoss', loss_weight=1.0),
+                 rotation_agnostic_classes: list = None,
+                 weak_supervised: bool = True,
+                 square_classes: list = None,
+                 crop_size: Tuple[int, int] = (768, 768),
+                 **kwargs):
```
Comment (zytx121) [inline mmrotate/models/dense_heads/h2rbox_head.py:93]:
```suggestion
                 **kwargs) -> None:
```

File: mmrotate/models/dense_heads/h2rbox_head.py [hunk 8]
```diff
@@ -0,0 +1,710 @@
...
+            loss_angle=loss_angle,
+            **kwargs)
+
+        self.loss_bbox_ss = MODELS.build(loss_bbox_ss)
+        self.rotation_agnostic_classes = rotation_agnostic_classes
+        self.weak_supervised = weak_supervised
+        self.square_classes = square_classes
+        self.crop_size = crop_size
+
+    def obb2xyxy(self, rbboxes):
```
Comment (zytx121) [inline mmrotate/models/dense_heads/h2rbox_head.py:115]:
```suggestion
    def obb2xyxy(self, rbboxes: Tensor) - > Tensor:
```

File: mmrotate/models/dense_heads/h2rbox_head.py [hunk 9]
```diff
@@ -0,0 +1,710 @@
...
+        dy = rbboxes[..., 1]
+        dw = hbbox_w.reshape(-1)
+        dh = hbbox_h.reshape(-1)
+        x1 = dx - dw / 2
+        y1 = dy - dh / 2
+        x2 = dx + dw / 2
+        y2 = dy + dh / 2
+        return torch.stack((x1, y1, x2, y2), -1)
+
+    def _process_rotation_agnostic(self, tensor, cls, dim=4):
```
Comment (zytx121) [inline mmrotate/models/dense_heads/h2rbox_head.py:133]:
```suggestion
    def _process_rotation_agnostic(self, tensor: Tensor, cls: Tensor, dim: int = 4) -> Tensor:
```

File: mmrotate/models/losses/h2rbox_consistency_loss.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    def __init__(self,
+                 center_loss_cfg: ConfigType = dict(
+                     type='mmdet.L1Loss', loss_weight=0.0),
+                 shape_loss_cfg: ConfigType = dict(
+                     type='mmdet.IoULoss', loss_weight=1.0),
+                 angle_loss_cfg: ConfigType = dict(
+                     type='mmdet.L1Loss', loss_weight=1.0),
+                 reduction: str = 'mean',
+                 loss_weight: float = 1.0) -> None:
+        super(H2RBoxConsistencyLoss, self).__init__()
```
Comment (zytx121) [inline mmrotate/models/losses/h2rbox_consistency_loss.py:20]:
```suggestion
        super().__init__()
```

File: mmrotate/models/losses/h2rbox_consistency_loss.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+        self.center_loss = MODELS.build(center_loss_cfg)
+        self.shape_loss = MODELS.build(shape_loss_cfg)
+        self.angle_loss = MODELS.build(angle_loss_cfg)
+        self.reduction = reduction
+        self.loss_weight = loss_weight
+
+    def forward(self,
+                pred,
```
Comment (zytx121) [inline mmrotate/models/losses/h2rbox_consistency_loss.py:28]:
```suggestion
                pred: Tensor,
```

File: mmrotate/models/losses/h2rbox_consistency_loss.py [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+                target,
```
Comment (zytx121) [inline mmrotate/models/losses/h2rbox_consistency_loss.py:29]:
```suggestion
                target: Tensor,
```

File: mmrotate/models/losses/h2rbox_consistency_loss.py [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+                weight,
```
Comment (zytx121) [inline mmrotate/models/losses/h2rbox_consistency_loss.py:30]:
```suggestion
                weight: Tensor,
```

File: mmrotate/models/losses/h2rbox_consistency_loss.py [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+                avg_factor=None,
```
Comment (zytx121) [inline mmrotate/models/losses/h2rbox_consistency_loss.py:31]:
```suggestion
                avg_factor: Optional[int] =None,
```

File: mmrotate/models/losses/h2rbox_consistency_loss.py [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+                reduction_override=None):
```
Comment (zytx121) [inline mmrotate/models/losses/h2rbox_consistency_loss.py:32]:
```suggestion
                reduction_override: Optional[str] = None) -> Tensor:
```

File: mmrotate/models/losses/h2rbox_consistency_loss.py [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+            target (Tensor): Corresponding gt boxes.
+            weight (Tensor): The weight of loss for each prediction.
+            avg_factor (int, optional): Average factor that is used to average
+                the loss. Defaults to None.
+            reduction_override (str, optional): The reduction method used to
+               override the original reduction method of the loss.
+               Defaults to None.
+
+        Returns:
+            Calculated loss (torch.Tensor)
```
Comment (zytx121) [inline mmrotate/models/losses/h2rbox_consistency_loss.py:46]:
```suggestion
            Calculated loss (Tensor)
```

**Review by zytx121 — APPROVED** [review]
LGTM

**Review by zytx121 — DISMISSED** [review]

**Review by liuyanyi — APPROVED** [review]
LGTM

**Review by zytx121 — APPROVED** [review]

**Review by RangiLyu — APPROVED** [review]

**Pull request merged**